indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -146,10 +146,13 @@
   private final RetryPolicyFactory retryPolicyFactory;
 
   @JsonIgnore
-  private List<IndexTask> indexTaskSpecs;
+  private final AppenderatorsManager appenderatorsManager;
 
   @JsonIgnore
-  private AppenderatorsManager appenderatorsManager;
+  private List<IndexTask> indexTaskSpecs;
+
+  @Nullable
+  private volatile IndexTask currentRunningTaskSpec = null;
```

Comment:
Added javadoc.

Reply:
The added Javadoc doesn't answer the question "Why the semantics of volatile field reads and writes (as defined in the Java Memory Model) are required for the field?" from the referenced checklist item.

It's actually likely that just adding `volatile` to the field is either not enough to achieve certain concurrency goal (which is also not clarified in the Javadoc), or not needed.

"by an HTTP thread when {@link #stopGracefully} is called." - please point to a specific class and method where does this happen, because it's not obvious.

Reply:
> The added Javadoc doesn't answer the question "Why the semantics of volatile field reads and writes (as defined in the Java Memory Model) are required for the field?" from the referenced checklist item.

I don't understand your comment. Do you want me to use the terms defined in the JMM?

> It's actually likely that just adding volatile to the field is either not enough to achieve certain concurrency goal (which is also not clarified in the Javadoc), or not needed.

I don't think this kind of comment helps. So do you think volatile here is necessary or not?

Reply:
> Do you want me to use the terms defined in the JMM?

Yes, I think the term "happens-before" should usually appear in this type of explanatory comments.

> So do you think volatile here is necessary or not?

I don't know because I wasn't able to figure out the place you referred to in "by an HTTP thread when `{@link #stopGracefully}` is called."

Reply:
But first of all, "This variable is updated by the main thread and read by an HTTP thread when `{@link #stopGracefully}` is called." doesn't explain the concurrency model in enough detail. Can `{@link #stopGracefully}` be called fully concurrently at whatever moment in parallel with `runTask()`, or it could be called only in response to some failure within `runTask()`, just from a different thread? In the first case, `volatile` is likely needed to ensure *safe publication* (see JCIP 3.5, or [here](https://shipilev.net/blog/2014/safe-public-construction/#_safe_publication)) of the `IndexTask` object. In the second case, volatile is almost definitely redundant, if the thread from which `stopGracefully()` is called "gets informed" about the failure in `runTask()` in any way that is different from something like a spin-loop read of a non-volatile field (yes, this is very obscure and never actually happens in practice).

Reply:
In this particular API, `stopGracefully` can be called at any point during `run`. In response, the Task should "arrange for its 'run' method to exit promptly". So I think the `volatile` is needed for safe publication. To me a comment like this would explain it sufficiently.

> The sub-task that is currently running. Volatile since it will potentially be accessed by `stopGracefully` concurrently with `runTask`, which is responsible for assigning the value.

Side note, I don't think this applies here, but in general do you think it's a valid explanation to say: "this field is accessed by multiple threads, and I haven't conclusively proven that volatile is _not_ necessary, so I've included it"?

Probably a lot of the volatiles in the codebase today are like this. I don't necessarily see a big problem with it. Even though some of them could probably be safely removed, the 'just-in-case' volatiles still help lower the cognitive overhead of documenting and dealing with the codebase. (Otherwise, you'd need to think through these arguments again every time you make a change.) And in most cases, including this one, there is negligible performance cost.

Reply:
@leventov @gianm thanks, updated javaodc for `currentRunningTaskSpec` and `stopGracefully()`.

Reply:
So if `stopGracefully()` may be called at any time concurrently with `run()`, I think the concurrency property that is worth ensuring is not starting running the next task if `stopGracefully()` is already stopping the previous task. To achieve this, `volatile` is not enough. This property could be achieved the following way:
 - `currentRunningTaskSpec` is `AtomicReference<Object>`
 - The stopping callback:
 ```java
 Object currentRunningTask = currentRunningTaskSpec.getAndSet(SPECIAL_VALUE_STOPPED);
 if (currentRunningTask != null) {
  ((IndexTask) currentRunningTask).stopGracefull(config);
 }
```
 - The code in the loop in `runTask()` does something like
 ```java
 
 Object prevSpec = currentRunningTaskSpec.get();
 if (prevTask == SPECIAL_VALUE_STOPPED ||
    !currentRunningTaskSpec.compareAndSet(prevSpec, eachSpec)) {
  log.info("Stopped concurrently");
  ...
}
```

Reply:
That construction definitely looks better!

Reply:
> Side note, I don't think this applies here, but in general do you think it's a valid explanation to say: "this field is accessed by multiple threads, and I haven't conclusively proven that volatile is not necessary, so I've included it"?
>
> Probably a lot of the volatiles in the codebase today are like this. I don't necessarily see a big problem with it. Even though some of them could probably be safely removed, the 'just-in-case' volatiles still help lower the cognitive overhead of documenting and dealing with the codebase. (Otherwise, you'd need to think through these arguments again every time you make a change.)

To me, the main problem is that `volatile` is often used to delude oneself or reviewers that some code is concurrently safe without forcing through the threading/concurrent control flow models of the code, the desired properties of the code, and the proof that the code has the desired properties. If "`volatile` = probably worse perf, but no concurrency defects" was the case, that wouldn't be too bad, indeed. Unfortunately, in reality, what we have is "`volatile` = high probability that there are some concurrency defects around the code".

Reply:
Thanks, that's a good point. Fixed as suggested.

Reply:
> Unfortunately, in reality, what we have is "volatile = high probability that there are some concurrency defects around the code".

Interesting point. I wonder if we can nail down what usages of volatile are especially likely to be buggy. My intuitions would be:

- Not very risky: 'one-way' volatiles that are used to ensure safe publishing out of an 'owner' thread, but where the 'non-owner' threads aren't intended to call any mutation methods on the published object. This kind of pattern is sometimes used for publishing objects that monitoring or status-checking code will periodically inspect.
- Not very risky: volatiles used for deferred initialization (updated once from null -> nonnull, but not updated ever again).
- More risky: designs where a reference is updated multiple times by an 'owner' thread, and 'non-owner' threads may call some mutation methods on the object.
- Even more risky: designs where there is no clear 'owner' thread.

(The original example of `currentRunningTaskSpec` would have seemed risky by the above intuitions: there is a clear owner thread, but the reference is updated more than once, and non-owners are going to be calling mutation methods)